lib/io/QuadSinkHdt.ts

```diff
@@ -0,0 +1,75 @@
+import * as fs from 'fs/promises';
+import * as Path from 'path';
+import * as readline from 'readline';
+import type * as RDF from '@rdfjs/types';
+import * as Docker from 'dockerode';
+import type { IQuadSinkFileOptions } from './QuadSinkFile';
+import { QuadSinkFile } from './QuadSinkFile';
+import { transformToHdt, pullHdtCppDockerImage } from './rfdhdtDockerUtil';
+
+export class QuadSinkHdt extends QuadSinkFile {
+  private readonly files: Set<string> = new Set();
+  private readonly deleteSourceFiles: boolean;
+  private readonly hdtConversionOpPoolSize: number;
+
+  public constructor(options: IQuadSinkFileOptions, hdtConversionOpPoolSize = 5, deleteSourceFiles = true) {
+    super(options);
+    this.deleteSourceFiles = deleteSourceFiles;
+    this.hdtConversionOpPoolSize = hdtConversionOpPoolSize;
+  }
+
+  public async push(iri: string, quad: RDF.Quad): Promise<void> {
+    const path = Path.join('./', this.getFilePath(iri));
+    await super.push(iri, quad);
+
+    // Add files with the defined extension to the list to be transformed
+    if (this.fileExtension !== undefined && path.includes(this.fileExtension.replace('$', ''))) {
+      this.files.add(path);
+    }
+  }
+
+  /**
+   * Log the number of files transformed into HDT
+   * @param {number} counter - counter of the number of files transformed
+   * @param {boolean} newLine - add a new line after the logging
+   */
+  private attemptLogHdtTransformation(counter: number, newLine = false): void {
+    if (this.log) {
+      readline.clearLine(process.stdout, 0);
+      readline.cursorTo(process.stdout, 0);
+      process.stdout.write(`\rfiles transformed to HDT:${counter} out of ${this.files.size}`);
+      if (newLine) {
+        process.stdout.write(`\n`);
+      }
+    }
+  }
+
+  public async close(): Promise<void> {
+    // Close the streaming of files
+    await super.close();
+    const docker: Docker = new Docker();
+    // Pull the docker image if it is not available in the system
+    await pullHdtCppDockerImage(docker);
+
+    let i = 0;
+    let pool: Promise<void>[] = [];
+    for (const file of this.files) {
+      pool.push(this.transformToHdt(docker, file));
+      if (pool.length === this.hdtConversionOpPoolSize) {
+        await Promise.all(pool);
```

Comment:
This pool management is not as efficient as it can be.
Every time this line is reached, _all_ entries in the pool will be awaited.
It would be better if another entry can get started as soon as another one is ready.
(but something like this would probably be better to implement in a separate generic `Pool` class)

This is not something that needs to be done in this PR as far as I'm concerned, but let's make an issue for it then.

Reply:
I didn't thought about it... I don't mind to do it now with this PR since it can be quite slow. Do you think [worker thread](https://nodejs.org/api/worker_threads.html) would be better in this situation?

Reply:
Worker threads could also help a bit, but will probably make the implementation _way_ more complex. So I wouldn't recommend it.